tests/testthat/test-observe.R

```diff
@@ -15,14 +15,18 @@ describe("metaObserve", isolate({
   it("basically works 2", {
     x <- 0
     mo <- metaObserve2({
+      x <<- x + 1
       metaExpr({
-        x <<- 1
+        x <<- ..(x + 1)
       })
     })
     shiny:::flushReact()
 
-    expect_identical(x, 1)
-    expect_equal(withMetaMode(mo()), quote( x <<- 1 ))
+    expect_identical(x, 2)
+
+    # The value becomes 3 here because even `withMetaMode(mo())` has a side effect
+    # of x <<- x + 1 (the part outside the metaExpr)
+    expect_equal(withMetaMode(mo()), quote( x <<- 3 ))
```

Comment:
Huh, weird, I get `x <<- 4` when I run the following code, which is actually what I expect, but maybe I'm missing something?

```r
x <- 0
mo <- metaObserve2({
  x <<- x + 1
  metaExpr({
    x <<- ..(x + 1)
  })
})
shiny:::flushReact()
withMetaMode(mo())
```

Reply:
Oh, you're totally right. There's a bug in `metaObserve2`; dang, these are so subtle. I'll fix this and add some more tests.

Reply:
Basically the problem was that in meta mode only, `metaObserve2` (and `metaRender2`, and `metaReactive2`) all failed to introduce their own scope; so in the code above, `x <<- x + 1` is happening in the same environment as `x <- 0`. If `x <- 0` is not in the global environment, then this is introducing a new variable `x` in the global environment, while leaving the `x` that we were trying to change, unchanged. But when you try that test in the global environment, then `x <<- x + 1` does have the desired effect, because the `x` is already in the global environment.